lakey_service/downloader/models.py

```diff
@@ -139,15 +145,28 @@ def estimate_size(self, spec, catalogue_item_id):
         filter_chunks = Chunk.objects.filter(
             catalogue_item_id=catalogue_item_id, **query)
 
-        size = 0
+        c_i_size_by_col = {col['name']: col['size'] for col in c_i.spec}
+        c_i_count_by_col = {
+            col['name']: sum(dist['count'] for dist in col['distribution'])
+            for col in c_i.spec
+        }
+        c_i_size = sum(size for size in c_i_size_by_col.values())
+
+        estimate_size = 0
         for chunk in filter_chunks:
             for border in chunk.borders:
-                for dist in border['distribution']:
-                    size += (
-                        dist['count'] *
-                        self.column_type_to_bit_size[border['type']])
+                border_count = \
+                    sum(dist['count'] for dist in border['distribution'])
+                c_i_col_size = c_i_size_by_col[border['column']]
+                c_i_col_count = c_i_count_by_col[border['column']]
+
+                estimate_size += (border_count * c_i_col_size) / c_i_col_count
+
+        if estimate_size == c_i_size:
+            raise TooMuchDataRequestDetected(
+                f"specify filters to a smaller data area")
 
-        return size
+        return {'size': estimate_size, 'chunks': filter_chunks}
```

Comment:
you don't have to return a dictionary. Tuple is perfectly fine and allows the unpacking:

```python
return estimated_size, filter_chunks

# -- and where the called calls it
size, chunks = DownloadRequest.objects.estimate_size(spec)
```